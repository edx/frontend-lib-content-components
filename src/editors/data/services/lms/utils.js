import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';

/**
 * get(url)
 * simple wrapper providing an authenticated Http client get action
 * @param {string} url - target url
 */
export const get = (...args) => getAuthenticatedHttpClient().get(...args);
/**
 * post(url, data)
 * simple wrapper providing an authenticated Http client post action
 * @param {string} url - target url
 * @param {object|string} data - post payload
 */
export const post = (...args) => getAuthenticatedHttpClient().post(...args);

export const client = getAuthenticatedHttpClient;
