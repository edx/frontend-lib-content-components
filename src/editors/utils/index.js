/* eslint-disable import/prefer-default-export */
export { keyStore, StrictDict } from '../../utils';
export { default as camelizeKeys } from './camelizeKeys';
