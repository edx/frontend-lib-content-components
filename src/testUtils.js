/**
 * Mocked formatMessage provided by react-intl
 */
export const formatMessage = (msg, values) => {
  let message = msg.defaultMessage;
  if (values === undefined) {
    return message;
  }
  Object.keys(values).forEach((key) => {
    // eslint-disable-next-line
    message = message.replace(`{${key}}`, values[key]);
  });
  return message;
};

/**
 * Mock a single component, or a nested component so that its children render nicely
 * in snapshots.
 * @param {string} name - parent component name
 * @param {obj} contents - object of child components with intended component
 *   render name.
 * @return {func} - mock component with nested children.
 *
 * usage:
 *   mockNestedComponent('Card', { Body: 'Card.Body', Form: { Control: { Feedback: 'Form.Control.Feedback' }}... });
 *   mockNestedComponent('IconButton', 'IconButton');
 */
export const mockNestedComponent = (name, contents) => {
  if (typeof contents !== 'object') {
    return contents;
  }
  const fn = () => name;
  Object.defineProperty(fn, 'name', { value: name });
  Object.keys(contents).forEach((nestedName) => {
    const value = contents[nestedName];
    fn[nestedName] = typeof value !== 'object'
      ? value
      : mockNestedComponent(`${name}.${nestedName}`, value);
  });
  return fn;
};

/**
 * Mock a module of components.  nested components will be rendered nicely in snapshots.
 * @param {obj} mapping - component module mock config.
 * @return {obj} - module of flat and nested components that will render nicely in snapshots.
 * usage:
 *   mockNestedComponents({
 *     Card: { Body: 'Card.Body' },
 *     IconButton: 'IconButton',
 *   })
 */
export const mockNestedComponents = (mapping) => Object.entries(mapping).reduce(
  (obj, [name, value]) => ({
    ...obj,
    [name]: mockNestedComponent(name, value),
  }),
  {},
);
