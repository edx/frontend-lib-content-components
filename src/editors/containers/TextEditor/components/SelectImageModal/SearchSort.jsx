import React from 'react';
import PropTypes from 'prop-types';

import {
  ActionRow, Dropdown, Form, Icon, IconButton,
} from '@edx/paragon';
import { Close, Search } from '@edx/paragon/icons';
import { FormattedMessage, injectIntl, intlShape } from '@edx/frontend-platform/i18n';

import { sortKeys, sortMessages } from './utils';
import messages from './messages';

export const SearchSort = ({
  searchString,
  onSearchChange,
  clearSearchString,
  sortBy,
  onSortClick,
  // injected
  intl,
}) => (
  <ActionRow>
    <Form.Group style={{ margin: 0 }}>
      <Form.Control
        autoFocus
        onChange={onSearchChange}
        placeholder={intl.formatMessage(messages.searchPlaceholder)}
        trailingElement={
            searchString
              ? (
                <IconButton
                  iconAs={Icon}
                  invertColors
                  isActive
                  onClick={clearSearchString}
                  size="sm"
                  src={Close}
                />
              )
              : <Icon src={Search} />
          }
        value={searchString}
      />
    </Form.Group>
    <ActionRow.Spacer />
    <Dropdown>
      <Dropdown.Toggle id="img-sort-button" variant="tertiary">
        <FormattedMessage {...sortMessages[sortBy]} />
      </Dropdown.Toggle>
      <Dropdown.Menu>
        {Object.keys(sortKeys).map(key => (
          <Dropdown.Item key={key} onClick={onSortClick(key)}>
            <FormattedMessage {...sortMessages[key]} />
          </Dropdown.Item>
        ))}
      </Dropdown.Menu>
    </Dropdown>
  </ActionRow>
);

SearchSort.propTypes = {
  searchString: PropTypes.string.isRequired,
  onSearchChange: PropTypes.func.isRequired,
  clearSearchString: PropTypes.func.isRequired,
  sortBy: PropTypes.string.isRequired,
  onSortClick: PropTypes.func.isRequired,
  // injected
  intl: intlShape.isRequired,
};

export default injectIntl(SearchSort);
