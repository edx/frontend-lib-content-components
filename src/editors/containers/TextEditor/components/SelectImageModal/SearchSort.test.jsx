import React from 'react';
import { shallow } from 'enzyme';
import { Dropdown } from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import { formatMessage } from '../../../../../testUtils';

import { sortKeys, sortMessages } from './utils';
import { SearchSort } from './SearchSort';

describe('SearchSort component', () => {
  const props = {
    searchString: 'props.searchString',
    onSearchChange: jest.fn().mockName('props.onSearchChange'),
    clearSearchString: jest.fn().mockName('props.clearSearchString'),
    sortBy: sortKeys.dateOldest,
    onSortClick: jest.fn().mockName('props.onSortClick'),
    intl: { formatMessage },
  };
  describe('snapshots', () => {
    test('with search string (close button)', () => {
      expect(shallow(<SearchSort {...props} />)).toMatchSnapshot();
    });
    test('without search string (search icon)', () => {
      expect(shallow(<SearchSort {...props} searchString="" />)).toMatchSnapshot();
    });
    test('adds a sort option for each sortKey', () => {
      const el = shallow(<SearchSort {...props} />);
      expect(el.find(Dropdown).containsMatchingElement(
        <FormattedMessage {...sortMessages.dateNewest} />,
      )).toEqual(true);
      expect(el.find(Dropdown).containsMatchingElement(
        <FormattedMessage {...sortMessages.dateOldest} />,
      )).toEqual(true);
      expect(el.find(Dropdown).containsMatchingElement(
        <FormattedMessage {...sortMessages.nameAscending} />,
      )).toEqual(true);
      expect(el.find(Dropdown).containsMatchingElement(
        <FormattedMessage {...sortMessages.nameDescending} />,
      )).toEqual(true);
    });
  });
});
