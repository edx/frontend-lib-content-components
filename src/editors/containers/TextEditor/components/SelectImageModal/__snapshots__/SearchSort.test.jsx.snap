// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchSort component snapshots with search string (close button) 1`] = `
<ActionRow>
  <Form.Group
    style={
      Object {
        "margin": 0,
      }
    }
  >
    <Form.Control
      autoFocus={true}
      onChange={[MockFunction props.onSearchChange]}
      placeholder="Search"
      trailingElement={
        <IconButton
          iconAs="Icon"
          invertColors={true}
          isActive={true}
          onClick={[MockFunction props.clearSearchString]}
          size="sm"
          src={[MockFunction icons.Close]}
        />
      }
      value="props.searchString"
    />
  </Form.Group>
  <ActionRow.Spacer />
  <Dropdown>
    <Dropdown.Toggle
      id="img-sort-button"
      variant="tertiary"
    >
      <FormattedMessage
        defaultMessage="By date added (oldest)"
        description="Dropdown label for sorting by date (oldest)"
        id="authoring.texteditor.selectimagemodal.sort.dateoldest.label"
      />
    </Dropdown.Toggle>
    <Dropdown.Menu>
      <Dropdown.Item
        key="dateNewest"
      >
        <FormattedMessage
          defaultMessage="By date added (newest)"
          description="Dropdown label for sorting by date (newest)"
          id="authoring.texteditor.selectimagemodal.sort.datenewest.label"
        />
      </Dropdown.Item>
      <Dropdown.Item
        key="dateOldest"
      >
        <FormattedMessage
          defaultMessage="By date added (oldest)"
          description="Dropdown label for sorting by date (oldest)"
          id="authoring.texteditor.selectimagemodal.sort.dateoldest.label"
        />
      </Dropdown.Item>
      <Dropdown.Item
        key="nameAscending"
      >
        <FormattedMessage
          defaultMessage="By name (ascending)"
          description="Dropdown label for sorting by name (ascending)"
          id="authoring.texteditor.selectimagemodal.sort.nameascending.label"
        />
      </Dropdown.Item>
      <Dropdown.Item
        key="nameDescending"
      >
        <FormattedMessage
          defaultMessage="By name (descending)"
          description="Dropdown label for sorting by name (descending)"
          id="authoring.texteditor.selectimagemodal.sort.namedescending.label"
        />
      </Dropdown.Item>
    </Dropdown.Menu>
  </Dropdown>
</ActionRow>
`;

exports[`SearchSort component snapshots without search string (search icon) 1`] = `
<ActionRow>
  <Form.Group
    style={
      Object {
        "margin": 0,
      }
    }
  >
    <Form.Control
      autoFocus={true}
      onChange={[MockFunction props.onSearchChange]}
      placeholder="Search"
      trailingElement={<Icon />}
      value=""
    />
  </Form.Group>
  <ActionRow.Spacer />
  <Dropdown>
    <Dropdown.Toggle
      id="img-sort-button"
      variant="tertiary"
    >
      <FormattedMessage
        defaultMessage="By date added (oldest)"
        description="Dropdown label for sorting by date (oldest)"
        id="authoring.texteditor.selectimagemodal.sort.dateoldest.label"
      />
    </Dropdown.Toggle>
    <Dropdown.Menu>
      <Dropdown.Item
        key="dateNewest"
      >
        <FormattedMessage
          defaultMessage="By date added (newest)"
          description="Dropdown label for sorting by date (newest)"
          id="authoring.texteditor.selectimagemodal.sort.datenewest.label"
        />
      </Dropdown.Item>
      <Dropdown.Item
        key="dateOldest"
      >
        <FormattedMessage
          defaultMessage="By date added (oldest)"
          description="Dropdown label for sorting by date (oldest)"
          id="authoring.texteditor.selectimagemodal.sort.dateoldest.label"
        />
      </Dropdown.Item>
      <Dropdown.Item
        key="nameAscending"
      >
        <FormattedMessage
          defaultMessage="By name (ascending)"
          description="Dropdown label for sorting by name (ascending)"
          id="authoring.texteditor.selectimagemodal.sort.nameascending.label"
        />
      </Dropdown.Item>
      <Dropdown.Item
        key="nameDescending"
      >
        <FormattedMessage
          defaultMessage="By name (descending)"
          description="Dropdown label for sorting by name (descending)"
          id="authoring.texteditor.selectimagemodal.sort.namedescending.label"
        />
      </Dropdown.Item>
    </Dropdown.Menu>
  </Dropdown>
</ActionRow>
`;
